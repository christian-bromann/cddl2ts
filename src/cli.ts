import fs from 'node:fs/promises'
import path from 'node:path'

import { parse } from 'cddl'

import { transform } from './index.js'
import { pkg } from './constants.js'

const HELP = `
${pkg.name}
${pkg.description}

Usage:
runme2ts ./path/to/spec.cddl &> ./path/to/interface.ts

v${pkg.version}
Copyright ${(new Date()).getFullYear()} ${pkg.author}
`

export default async function cli (args = process.argv.slice(2)) {
    if (args.includes('--help') || args.length === 0) {
        console.log(HELP);
        return process.exit(0)
    }
    if (args.includes('--version') || args.includes('-v')) {
        console.log(pkg.version);
        return process.exit(0)
    }

    const absoluteFilePath = path.resolve(process.cwd(), args[0])
    const hasAccess = await fs.access(absoluteFilePath).then(() => true, () => false)
    if (!hasAccess) {
        console.error(`Couldn't find or access source CDDL file at "${absoluteFilePath}"`)
        return process.exit(1)
    }

    const ast = parse(absoluteFilePath)
    console.log(transform(ast))
}
