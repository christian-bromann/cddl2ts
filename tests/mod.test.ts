import url from 'node:url'
import path from 'node:path'
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'

import cli from '../src/cli.js'

const __dirname = url.fileURLToPath(new URL('.', import.meta.url))
const tsCDDL = path.join(__dirname, '..', '__fixtures__', 'test.cddl')

vi.mock('../src/constants', () => ({
    pkg: {
        name: 'foobar',
        version: '1.2.3'
    }
}))

describe('cddl2ts', () => {
    const exitOrig = process.exit.bind(process)
    const logOrig = console.log.bind(console)
    const errorOrig = console.error.bind(console)
    beforeEach(() => {
        process.exit = vi.fn()
        console.log = vi.fn()
        console.error = vi.fn()
    })

    it('should print help if no args were provided', async () => {
        await cli([])
        expect(console.log).toBeCalledWith(expect.stringMatching(/foobar/))
        expect(process.exit).toBeCalledWith(0)
    })

    it('should print help --help is set in args', async () => {
        await cli(['foo', 'bar', '--help', 'barfoo'])
        expect(console.log).toBeCalledWith(expect.stringMatching(/foobar/))
        expect(process.exit).toBeCalledWith(0)
    })

    it('should print version', async () => {
        await cli(['foo', '-v', 'bar'])
        expect(console.log).toBeCalledWith('1.2.3')
        expect(process.exit).toBeCalledWith(0)
    })

    it('should print version', async () => {
        await cli(['foo', '--version', 'bar'])
        expect(console.log).toBeCalledWith('1.2.3')
        expect(process.exit).toBeCalledWith(0)
    })

    it('should fail if first parameter is not pointing to a file', async () => {
        await cli(['foo'])
        expect(console.error).toBeCalledTimes(1)
        expect(process.exit).toBeCalledWith(1)
    })

    it('should fail if first parameter is not pointing to a file', async () => {
        await cli([path.join(__dirname, '__fixtures__', 'test.cddl')])
        expect(vi.mocked(console.log).mock.calls).toMatchSnapshot()
        expect(process.exit).toBeCalledTimes(0)
    })

    afterEach(() => {
        process.exit = exitOrig
        console.log = logOrig
        console.error = errorOrig
    })
})
