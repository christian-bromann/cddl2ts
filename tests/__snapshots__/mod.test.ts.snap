// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cddl2ts > should fail if first parameter is not pointing to a file 1`] = `
[
  [
    "export type Message = CommandResponse | ErrorResponse | Event;

export interface ErrorResponse extends Extensible {
  id?: JsUint | null;
  error: ErrorCode;
  message: string;
  stacktrace?: string;
}

export type ResultData = BrowsingContextResult | EmptyResult | NetworkResult | ScriptResult | SessionResult;
export type ErrorCode = \\"invalid argument\\" | \\"invalid session id\\" | \\"no such alert\\" | \\"no such frame\\" | \\"no such handle\\" | \\"no such node\\" | \\"no such script\\" | \\"session not created\\" | \\"unknown command\\" | \\"unknown error\\" | \\"unsupported operation\\";
export type JsInt = number;
export type JsUint = number;
export type Handle = string;
export type SharedId = string;

export interface UndefinedValue {
  type: \\"undefined\\";
}

export interface NullValue {
  type: null;
}

export interface StringValue {
  type: \\"string\\";
  value: string;
}

export type SpecialNumber = \\"NaN\\" | \\"-0\\" | \\"Infinity\\" | \\"-Infinity\\";

export interface NumberValue {
  type: \\"number\\";
  value: Number | SpecialNumber;
}

export interface SessionCapabilitiesRequest {
  alwaysMatch?: SessionCapabilityRequest;
  firstMatch?: SessionCapabilityRequest[];
  nested: {
    obj: 123;
    moreNesting: {
      id: JsUint | null;
      like: boolean;
    };
  };
}

export interface SomeGroup {
  /**
   * this is the optional leading param
   * this is not leading
   *
   * @default 'foobar'
   */
  optional?: string;
  /**
   * @default 'portrait'
   */
  orientation?: \\"portrait\\" | \\"landscape\\";
  /**
   * @default 1
   */
  scale?: number;
  /**
   * @default true
   */
  shrinkToFit?: boolean;
  /**
   * @default true
   */
  shrinkToFitP?: boolean;
  /**
   * this is the pageRanges param
   *
   * @default 1
   */
  bottom?: number;
  pageRanges?: (JsUint | string)[];
  /**
   * @default null
   */
  foo?: string;
  /**
   * @default 'mouse'
   */
  pointerType?: InputPointerType;
}

export type ScriptListLocalValue = (ScriptLocalValue)[];
export type ScriptMappingLocalValue = (ScriptLocalValue | ScriptLocalValue)[];

export // some comments here
type Extensible = Record<string, any>;",
  ],
]
`;
